jsonl-store/jsonl.ts

```diff
@@ -0,0 +1,242 @@
+import { JSONLinesParseStream } from "https://deno.land/x/jsonlines@v1.2.2/mod.ts";
+import { emptyDir, exists, walk } from "jsr:@std/fs@1.0.4";
+import {
+  basename,
+  dirname,
+  globToRegExp,
+  join,
+  toFileUrl,
+} from "jsr:@std/path@1.0.6";
+import {
+  call,
+  createQueue,
+  each,
+  type Operation,
+  resource,
+  spawn,
+  type Stream,
+  stream,
+} from "npm:effection@4.0.0-alpha.3";
+import type { Store, StoreConstructorOptions } from "./types.ts";
+
+import fs from "node:fs";
+import { promisify } from "node:util";
+
+function* mkdir(
+  path: fs.PathLike,
+  options: fs.MakeDirectoryOptions & {
+    recursive: true;
+  },
+): Operation<string | undefined> {
+  return yield* call(() => promisify(fs.mkdir)(path, options));
+}
+
+export class JSONLStore implements Store {
+  constructor(public location: URL) {}
+
+  /**
+   * Creates a store with a location that has a trailing slash.
+   * The trailing slash is important to ensure that the store content
+   * is written to the store directory and not the directory above which
+   * can be very annoying. The location has to be absolute.
+   * ```ts
+   * const store = JSONLStore.from({ location: 'file:///Users/foo/.store/' })
+   * ```
+   *
+   * @param options StoreConstructorOptions
+   * @returns
+   */
+  static from(options: StoreConstructorOptions): JSONLStore {
+    return new JSONLStore(
+      toFileUrl(`${options.location}`.replace(/\/?$/, "/")),
+    );
+  }
+
+  /**
+   * Returns true when key is present
+   * ```ts
+   * import { useStore } from "jsr:@effection-contrib/jsonl-store";
+   *
+   * const store = yield* useStore();
+   *
+   * if (yield* store.has("test")) {
+   *  console.log("store exists");
+   * }
+   * ```
+   *
+   * @param key string
+   * @returns boolean
+   */
+  *has(key: string): Operation<boolean> {
+    const location = new URL(`./${key}.jsonl`, this.location);
+
+    return yield* call(async () => {
+      try {
+        return await exists(location);
+      } catch {
+        return false;
+      }
+    });
+  }
+
+  /**
+   * Returns content of a file as a stream
+   *
+   * ```ts
+   * import { each } from "npm:effection@4.0.0-alpha.3";
+   * import { useStore } from "jsr:@effection-contrib/jsonl-store";
+   *
+   * const store = yield* useStore();
+   *
+   * for (const item of yield* each(store.read<number>("test"))) {
+   *   console.log(item)
+   *   yield* each.next();
+   * }
+   * ```
+   *
+   * @param key string
+   * @returns Stream<T>
+   */
+  read<T>(key: string): Stream<T, void> {
+    const location = new URL(`./${key}.jsonl`, this.location);
+
+    return resource(function* (provide) {
+      const file = yield* call(() => Deno.open(location, { read: true }));
```

Comment:
don't you need to close this file after you have opened it?

Reply:
No, it's unfortunately closed implicitly by the stream processor.